libpromises/lastseen.c

```diff
@@ -766,3 +766,46 @@ int RemoveKeysFromLastSeen(const char *input, bool must_be_coherent,
 
     return 0;
 }
+
+static bool OnlyRewriteIfChanged(KeyHostSeen *entry, ARG_UNUSED size_t size, KeyHostSeen *new)
+{
+    assert(entry != NULL);
+    assert(new != NULL);
+
+    if (entry->acknowledged)
+    {
+        return false;
+    }
+
+    new->acknowledged = true;
+    new->lastseen = entry->lastseen;
+    new->Q = entry->Q;
+
+    return true;
+}
+
+bool LastSeenHostAcknowledge(const char *host_key, bool incoming)
+{
+    DBHandle *db = NULL;
+    if (!OpenDB(&db, dbid_lastseen))
+    {
+        Log(LOG_LEVEL_ERR, "Unable to open lastseen DB");
+        return false;
+    }
+
+    // Update quality-of-connection entry
+    char key[CF_BUFSIZE];
+    NDEBUG_UNUSED int ret = snprintf(key, CF_BUFSIZE, "q%c%s", incoming ? 'i' : 'o', host_key);
+    assert(ret > 0 && ret < CF_BUFSIZE);
+
+    KeyHostSeen value;
+    if (!OverwriteDB(db, key, &value, sizeof(value), (OverwriteCondition)OnlyRewriteIfChanged, &value))
+    {
+        Log(LOG_LEVEL_ERR, "Unable to overwrite key '%s' to lastseen DB", key);
```

Comment:
Isn't this going to print the error for every host that is attempted to be acknowledged again which should be handled gracefully as a noop?

Reply:
Yes, I think you are right :+1: